File: python/823. Binary Trees With Factors.py
class Solution(object):
    def numFactoredBinaryTrees(self, A):
        """
        :type A: List[int]
        :rtype: int
        """
        mod = 10**9+7
        hash = collections.defaultdict(int)
        dp = collections.defaultdict(int)
        for i in A:
            hash[i] += 1
            dp[i] = 1
        n = len(A)
        res = 0
        A.sort()
        for k in range(n):
            cur = A[k]
            for j in range(k):
                if cur%A[j]==0 and cur/A[j] in hash:
                    dp[cur] = (dp[cur] + dp[A[j]] * dp[cur/A[j]]) % mod
            res = (res+dp[cur])%mod
        return res

File: go/src/leetcode/823/823. Binary Trees With Factors.go
package main

import "sort"

/*
*
https://leetcode.cn/problems/binary-trees-with-factors/description/
*/
const MOD int = 1e9 + 7

func numFactoredBinaryTrees(arr []int) (tot int) {
	f := make(map[int]int)
	d := make(map[int]int)
	for _, i := range arr {
		d[i]++
		f[i] = 1
	}
	sort.Ints(arr)
	for i, cur := range arr {
		for j := 0; j < i; j++ {
			if cur%arr[j] == 0 && d[cur/arr[j]] != 0 {
				f[cur] = (f[cur] + f[arr[j]]*f[cur/arr[j]]) % MOD
			}
		}
		tot = (tot + f[cur]) % MOD
	}
	return
}
